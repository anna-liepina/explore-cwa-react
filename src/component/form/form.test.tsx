import React from 'react';
import '@testing-library/jest-dom';
import { render, fireEvent, act } from '@testing-library/react';
import Form, { IFormSectionConfig } from './form';

describe('<Form/>', () => {
    const ctrlProps = {
        updateCTRL: {
            label: 'updateCTRL.label',
        },
        submitCTRL: {
            label: 'submitCTRL.label',
        },
        cancelCTRL: {
            label: 'cancelCTRL.cancel',
        },
    };

    const component = (props: React.ComponentPropsWithRef<'input'>) => <input {...props} />

    const props = {
        config: [
            {
                title: 'props.config[0].title',
                isCollapsed: false,
                items: [
                    {
                        component,
                        value: '',
                    }
                ]
            }
        ],
        ...ctrlProps,
    };

    const data = {};
    const optionalProps = {
        'data-cy': 'optProps.data-cy',
        className: 'optProps.className',
        title: 'optProps.title',
        data,
    };

    const e = {
        target: {
            value: '{{value}}',
            getAttribute: () => 0,
        },
    };

    describe('render', () => {
        it('with default/required props', () => {
            const { asFragment } = render(<Form {...props} />);

            expect(asFragment()).toMatchSnapshot();
        });

        it('with optional/required props', () => {
            const { asFragment } = render(<Form {...props} {...optionalProps} />);

            expect(asFragment()).toMatchSnapshot();
        });
    });

    describe('lifecycle events', () => {
        describe('::componentDidMount', () => {
            it('should invoke external callback [::onMount] as soon as component been mounted', async () => {
                const spy = jest.fn(async () => Promise.resolve({ config: [] }));

                await act(async () => {
                    render(<Form {...props} onMount={spy} />);
                });

                expect(spy).toBeCalledWith(
                    expect.any(Object),
                    expect.any(Object),
                );
            });
        });
    });

    describe('external callbacks', () => {
        describe('::onSuccess', () => {
            it('should invoke external callback [::onSuccess] if external callback [::onMount] succeed', async () => {
                const spy = jest.fn();
                const onMount = jest.fn(async () => Promise.resolve({ config: [] }));

                await act(async () => {
                    render(<Form {...props} onSuccess={spy} onMount={onMount} />);
                });

                expect(spy).toBeCalledWith(
                    expect.any(Object),
                    expect.any(Object)
                );
            });
        });

        describe('::onError', () => {
            it('should invoke external callback [::onError] if external callback [::onMount] fails', async () => {
                const spy = jest.fn();
                const onMount = jest.fn(async () => Promise.reject());

                await act(async() => {
                    render(<Form {...props} onError={spy} onMount={onMount} />);
                });

                expect(spy).toBeCalledWith(
                    expect.any(Object),
                    expect.any(Object),
                );
            });
        });

        describe('::onSubmit', () => {
            it('should NOT invoke external callback [::onSubmit] from click on [data-cy="form--submit"] because ::isValid is falsy', () => {
                const spy = jest.fn();
                const { container } = render(<Form {...props} onSubmit={spy} />);

                fireEvent.click(container.querySelector('[data-cy="form--submit"]')!);

                expect(spy).not.toBeCalled();
            });

            it('should invoke external callback [::onSubmit] from click on [data-cy="form--submit"]', async () => {
                const spy = jest.fn(async () => Promise.resolve({ config: [] }));
                const { container } = render(<Form {...props} onSubmit={spy} isValid />);

                await act(async () => {
                    fireEvent.click(container.querySelector('[data-cy="form--submit"]')!);
                });

                expect(spy).toBeCalledWith(
                    expect.any(Object),
                    expect.any(Object),
                );
            });

            it('should NOT invoke external callback [::onSubmit] from click on [data-cy="form--update"] because ::isValid is falsy', () => {
                const spy = jest.fn();
                const { container } = render(<Form {...props} onSubmit={spy} data={data} />);

                fireEvent.click(container.querySelector('[data-cy="form--update"]')!);

                expect(spy).not.toBeCalled();
            });

            it('should invoke external callback [::onSubmit] from click on [data-cy="form--update"]', async () => {
                const spy = jest.fn(async () => Promise.resolve({ config: [] }));
                const { container } = render(<Form {...props} onSubmit={spy} data={data} isValid />);

                await act(async () => {
                    fireEvent.click(container.querySelector('[data-cy="form--update"]')!);
                });

                expect(spy).toBeCalledWith(
                    expect.any(Object),
                    expect.any(Object),
                );
            });

            it('should invoke external callback [::validate] from click on [data-cy="form--submit"]', () => {
                const spy = jest.fn();
                const { container } = render(<Form {...props} validate={spy} isValid />);

                fireEvent.click(container.querySelector('[data-cy="form--submit"]')!);

                expect(spy).toBeCalledWith(props.config);
            });

            it('when external callback [::validate] return false, should not invoke external callback [::onSubmit]', () => {
                const spy = jest.fn();

                const { container } = render(<Form {...props}  onSubmit={spy} isValid validate={() => false} />);

                fireEvent.click(container.querySelector('[data-cy="form--submit"]')!);

                expect(spy).not.toBeCalled();
            });
        });

        describe('::onCancel', () => {
            it('should invoke external callback [::onCancel] from click on [data-cy="form--cancel"]', () => {
                const spy = jest.fn();
                const { container } = render(<Form {...props} onCancel={spy} />);

                fireEvent.click(container.querySelector('[data-cy="form--cancel"]')!);

                expect(spy).toBeCalledWith(
                    expect.any(Object),
                    expect.any(Object),
                );
            });
        });

        describe('::onCollapse', () => {
            it('from a click on <Accordion/> title, it should toggle ::isCollapsed, if ::isCollapsed truthy do not render <Accordion /> content', () => {
                const { container } = render(<Form {...props} />);

                expect(container.querySelector('[data-cy="section-0-input-0"')!).toBeInTheDocument();

                fireEvent.click(container.querySelector('[data-cy="section-0"]')!);

                expect(container.querySelector('[data-cy="section-0-input-0"')).not.toBeInTheDocument();
            });
        });

        describe('::onChange', () => {
            it('should invoke external callback [::validate] with section/field IDs', () => {
                const spy = jest.fn();
                props.config[0].isCollapsed = false;

                const { container } = render(<Form {...props} isValid validate={spy} />);

                fireEvent.change(container.querySelector('[data-cy="section-0-input-0"]')!, e);

                expect(spy).toBeCalledWith(props.config, [[0, 0]]);
            });

            it('should mutate state field [::config] and set [::value] for relevant field from payload', () => {
                props.config[0].isCollapsed = false;
                props.config[0].items = [
                    {
                        component,
                        value: '',
                    },
                    {
                        component,
                        value: '',
                    },
                ];

                const { container } = render(<Form {...props} isValid />);

                fireEvent.change(container.querySelector('[data-cy="section-0-input-0"]')!, e);

                expect(props.config).toMatchSnapshot();
            });

            it('should keep errors if validation engine is NOT provided', () => {
                const config = [
                    {
                        isCollapsed: false,
                        items: [
                            {
                                component,
                                value: '',
                            },
                            {
                                component,
                                value: '',
                                errors: ['{{error}}', '{{ another error }}'],
                            },
                        ],
                    },
                ];

                const { container } = render(<Form {...props} config={config} />);

                expect(container.querySelector('[errors]')!).toBeInTheDocument();

                fireEvent.change(container.querySelector('[data-cy="section-0-input-0"]')!, e);

                expect(container.querySelector('[errors]')!).toBeInTheDocument();
            });

            it('should NOT keep errors if validation was successful', () => {
                const removeErrors = (config: IFormSectionConfig[]) => {
                    config.forEach((section) => {
                        section.items.forEach((item) => {
                            delete item.errors;
                        })
                    })

                    return true;
                }

                const config = [
                    {
                        isCollapsed: false,
                        items: [
                            {
                                component,
                                value: '',
                            },
                            {
                                component,
                                value: '',
                                errors: ['{{error}}', '{{ another error }}'],
                            },
                        ],
                    },
                ];

                const { container } = render(<Form {...props} validate={removeErrors} config={config} />);

                expect(container.querySelector('[errors]')!).toBeInTheDocument();

                fireEvent.change(container.querySelector('[data-cy="section-0-input-0"]')!, e);

                expect(container.querySelector('[errors]')).not.toBeInTheDocument();
            });
        });
    });
});
