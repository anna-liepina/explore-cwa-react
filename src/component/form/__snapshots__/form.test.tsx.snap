// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form/> external callbacks ::onChange should mutate ::config and set ::value for relevant field from payload 1`] = `
Array [
  Object {
    "isCollapsed": false,
    "items": Array [
      Object {
        "component": [Function],
        "value": "{{value}}",
      },
      Object {
        "component": [Function],
        "value": "",
      },
    ],
    "title": "props.config[0].title",
  },
]
`;

exports[`<Form/> render with default/required props 1`] = `
<DocumentFragment>
  <form
    class="form"
  >
    <section
      class="accordion"
    >
      <h2
        class="accordion_title"
        data-cy="section-0"
        data-section="0"
      >
        props.config[0].title
      </h2>
      <input
        data-cy="section-0-input-0"
        data-field="0"
        data-section="0"
        value=""
      />
    </section>
    <div
      class="form_controls"
    >
      <button
        class="button"
        data-cy="form--submit"
        disabled=""
      >
        submitCTRL.label
      </button>
      <button
        class="button"
        data-cy="form--cancel"
      >
        cancelCTRL.cancel
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`<Form/> render with optional/required props 1`] = `
<DocumentFragment>
  <form
    class="form optProps.className"
  >
    <h1
      class="form_title"
      data-cy="optProps.data-cyform-title"
    >
      optProps.title
    </h1>
    <section
      class="accordion"
    >
      <h2
        class="accordion_title"
        data-cy="optProps.data-cysection-0"
        data-section="0"
      >
        props.config[0].title
      </h2>
      <input
        data-cy="optProps.data-cysection-0-input-0"
        data-field="0"
        data-section="0"
        value=""
      />
    </section>
    <div
      class="form_controls"
    >
      <button
        class="button"
        data-cy="optProps.data-cyform--update"
        disabled=""
      >
        updateCTRL.label
      </button>
      <button
        class="button"
        data-cy="optProps.data-cyform--cancel"
      >
        cancelCTRL.cancel
      </button>
    </div>
  </form>
</DocumentFragment>
`;
